src/main/java/net/minecraftforge/common/brewing/MixingBrewingRecipe.java

```diff
@@ -0,0 +1,115 @@
+import net.minecraft.world.item.Items;
+import net.minecraft.world.item.alchemy.Potion;
+import net.minecraft.world.item.alchemy.PotionUtils;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.item.crafting.RecipeSerializer;
+import net.minecraft.world.item.crafting.ShapedRecipe;
+import net.minecraftforge.registries.ForgeRegistries;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+import net.minecraftforge.registries.ObjectHolder;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public record MixingBrewingRecipe(@Nonnull ResourceLocation id,
+                                  @Nonnull Potion input,
+                                  @Nonnull Ingredient ingredient,
+                                  @Nonnull Potion output) implements IBrewingRecipe
+{
+    @ObjectHolder("forge:mixing_brewing")
+    public static final RecipeSerializer<MixingBrewingRecipe> SERIALIZER = null;
+
+    @Override
+    public boolean isInput(@Nonnull ItemStack stack)
+    {
+        final Item item = stack.getItem();
+        return (item == Items.POTION || item == Items.SPLASH_POTION || item == Items.LINGERING_POTION || item == Items.GLASS_BOTTLE) && PotionUtils.getPotion(stack) == input();
+    }
+
+    @Override
+    public ItemStack getOutput(ItemStack input, ItemStack ingredient)
+    {
+        return isInput(input) && isIngredient(ingredient) ? PotionUtils.setPotion(new ItemStack(input.getItem()), output()) : ItemStack.EMPTY;
+    }
+
+    @Override
+    public boolean isIngredient(ItemStack ingredient)
+    {
+        return this.ingredient.test(ingredient);
+    }
+
+    @Override
+    public ItemStack getResultItem()
+    {
+        return ItemStack.EMPTY;
+    }
+
+    @Override
+    public ResourceLocation getId()
+    {
+        return id;
+    }
+
+    @Override
+    public RecipeSerializer<?> getSerializer()
+    {
+        return SERIALIZER;
+    }
+
+    public static class Serializer extends ForgeRegistryEntry<RecipeSerializer<?>> implements RecipeSerializer<MixingBrewingRecipe>
+    {
+
+        @Override
+        public MixingBrewingRecipe fromJson(final ResourceLocation p_199425_1_, final JsonObject p_199425_2_)
+        {
+            Potion input = ForgeRegistries.POTION_TYPES.getValue(new ResourceLocation(p_199425_2_.get("input_potion").getAsString()));
+            Ingredient ingredient = Ingredient.fromJson(p_199425_2_.getAsJsonObject("ingredient"));
+            Potion output = ForgeRegistries.POTION_TYPES.getValue(new ResourceLocation(p_199425_2_.get("output_potion").getAsString()));
+            return new MixingBrewingRecipe(p_199425_1_, input, ingredient, output);
+        }
+
+        @Nullable
+        @Override
+        public MixingBrewingRecipe fromNetwork(final ResourceLocation p_199426_1_, final FriendlyByteBuf p_199426_2_)
+        {
+            Potion input = ForgeRegistries.POTION_TYPES.getValue(p_199426_2_.readResourceLocation());
```

Comment:
More unchecked NPEs.

Reply:
what should I throw in the network methods?

Reply:
Disregard this, I forgot that forge does infact default to `Potion#EMPTY` if the value is nonexistent which is vanilla behavior.